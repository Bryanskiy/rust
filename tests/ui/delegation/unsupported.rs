#![feature(c_variadic)]
#![feature(fn_delegation)]
#![allow(incomplete_features)]

mod generics {
    trait Trait0 {
        fn bar(&self) {}
    }

    trait Trait1<T> {
        fn foo(&self) {}
        reuse Trait0::bar;
        //~^ ERROR early bound generics are only supported for trait implementations and free functions
    }

    trait Trait2 {
        reuse Trait1::foo;
        //~^ ERROR early bound generics are only supported for trait implementations and free functions
    }

    struct S;
    impl S {
        reuse Trait1::foo;
        //~^ ERROR early bound generics are only supported for trait implementations and free functions
    }
}

mod opaque {
    trait Trait {}
    impl Trait for () {}

    mod to_reuse {
        use super::Trait;

        pub fn opaque_ret() -> impl Trait { () }
    }

    trait ToReuse {
        fn opaque_ret() -> impl Trait { () }
    }

    // FIXME: Inherited `impl Trait`s create query cycles when used inside trait impls.
    impl ToReuse for u8 {
        reuse to_reuse::opaque_ret; //~ ERROR cycle detected when computing type
    }
    impl ToReuse for u16 {
        reuse ToReuse::opaque_ret; //~ ERROR cycle detected when computing type
    }
}

mod recursive {
    mod to_reuse1 {
        pub mod to_reuse2 {
            pub fn foo() {}
        }

        pub reuse to_reuse2::foo;
    }

    reuse to_reuse1::foo;
    //~^ ERROR recursive delegation is not supported yet
}

fn main() {}
