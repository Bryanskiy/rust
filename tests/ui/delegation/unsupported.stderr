error[E0391]: cycle detected when computing type of `opaque::<impl at $DIR/unsupported.rs:20:5: 20:24>::{synthetic#0}`
  --> $DIR/unsupported.rs:21:25
   |
LL |         reuse to_reuse::opaque_ret;
   |                         ^^^^^^^^^^
   |
note: ...which requires comparing an impl and trait method signature, inferring any hidden `impl Trait` types in the process...
  --> $DIR/unsupported.rs:21:25
   |
LL |         reuse to_reuse::opaque_ret;
   |                         ^^^^^^^^^^
   = note: ...which again requires computing type of `opaque::<impl at $DIR/unsupported.rs:20:5: 20:24>::{synthetic#0}`, completing the cycle
note: cycle used when checking that `opaque::<impl at $DIR/unsupported.rs:20:5: 20:24>` is well-formed
  --> $DIR/unsupported.rs:20:5
   |
LL |     impl ToReuse for u8 {
   |     ^^^^^^^^^^^^^^^^^^^
   = note: see https://rustc-dev-guide.rust-lang.org/overview.html#queries and https://rustc-dev-guide.rust-lang.org/query.html for more information

error[E0391]: cycle detected when computing type of `opaque::<impl at $DIR/unsupported.rs:23:5: 23:25>::{synthetic#0}`
  --> $DIR/unsupported.rs:24:24
   |
LL |         reuse ToReuse::opaque_ret;
   |                        ^^^^^^^^^^
   |
note: ...which requires comparing an impl and trait method signature, inferring any hidden `impl Trait` types in the process...
  --> $DIR/unsupported.rs:24:24
   |
LL |         reuse ToReuse::opaque_ret;
   |                        ^^^^^^^^^^
   = note: ...which again requires computing type of `opaque::<impl at $DIR/unsupported.rs:23:5: 23:25>::{synthetic#0}`, completing the cycle
note: cycle used when checking that `opaque::<impl at $DIR/unsupported.rs:23:5: 23:25>` is well-formed
  --> $DIR/unsupported.rs:23:5
   |
LL |     impl ToReuse for u16 {
   |     ^^^^^^^^^^^^^^^^^^^^
   = note: see https://rustc-dev-guide.rust-lang.org/overview.html#queries and https://rustc-dev-guide.rust-lang.org/query.html for more information

error: recursive delegation is not supported yet
  --> $DIR/unsupported.rs:37:22
   |
LL |         pub reuse to_reuse2::foo;
   |                              --- callee defined here
...
LL |     reuse to_reuse1::foo;
   |                      ^^^

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0391`.
